casper/src/main/resources/PoS.rho

```diff
@@ -1,180 +1,165 @@
-                  )
-
-                  | contract isBonded (@userPk, return) = {
-                    for(@bonds <- bondsCh) {
-                      bondsCh!(bonds) | return!(bonds.contains(userPk))
-                    }
-                  }
-
-                  | contract validateBondAmount (@bondAmount, return) = {
-                    if (bondAmount < minimumBond) {
-                      return!((false, "Bond less than minimum!"))
-                    } else if (bondAmount > maximumBond) {
-                      return!((false, "Bond greater than maximum!"))
-                    } else {
-                      return!((true, bondAmount))
-                    }
-                  }
-
-                  // TODO: figure out rate-limiting of bonding
-                  | contract validateBondingRate (@bondAmount, return) = {
-                    return!((true, bondAmount))
-                  }
-
-                  | contract validatePublicKey (@bondAmount, return) = {
-                    new isBondedCh, parametersCh in {
-                      getParameters!(*parametersCh)
-                      | for (_, _, @userPk, _ <- parametersCh) {
-                        isBonded!(userPk, *isBondedCh)
-                        | for (@isBondedResult <- isBondedCh) {
-                          if (isBondedResult) {
-                            return!((false, "Public key ${pk} already bonded." %% {"pk" : userPk}))
-                          } else return!((true, bondAmount))
-                        }
-                      }
-                    }
-                  }
-
-                  | contract updateBonds (@bondAmount, return) = {
-                    new parametersCh in {
-                      getParameters!(*parametersCh)
-                      | for(@bonds <- bondsCh; _, _, @userPk, _ <- parametersCh) {
-                        bondsCh!(
-                          bonds.set(
-                            userPk,
-                            [bondAmount, bonds.size() + 1]
-                          )
-                        ) | return!((true, "Bond successful!"))
-                      }
+            } |
+
+            contract PoS (@"bond", @amount, returnCh) = {
+              new userCh, depositCh in {
+                getUser!(*userCh) |
+
+                for(@bonds <- bondsCh;
+                    @userPk <- userCh) {
+
+                  if (bonds.contains(userPk)) {
+                    bondsCh!(bonds) |
+                    returnCh!((false, "Public key is already bonded."))
+                  } else if (amount < minimumBond) {
+                    bondsCh!(bonds) |
+                    returnCh!((false, "Bond less than minimum!"))
+                  } else if (amount > maximumBond) {
+                    bondsCh!(bonds) |
+                    returnCh!((false, "Bond greater than maximum!"))
+                  } else {
+                    deposit!(userPk, amount, *depositCh) |
+                    for(@depositResult <- depositCh) {
```

Comment:
In this implementation, the bonds map is always updated, regardless of whether or not the deposit fails. You should only update the bonds map if the deposit succeeds.

Reply:
The bondsCh acts as a MVar so I take the map value, use it and then I put it back in the channel. If it changes I put back in the channel the updated value. If it stays the same I put back in the channel the original value.

If I don't put the value back even on errors the pos instance would become unusable after an error.

Reply:
I'll rephrase. You're returning an updated map to the `bondsCh` whether or not `deposit` succeeds. See the lines below. Explicitly, if deposit fails, you should return `bonds`. If it succeeds, you should should return what you're returning now to `bonds`.

Reply:
good point! I'll add a test for that case too